### PR TITLE
Improved time and date handling

### PR DESCRIPTION
**Pull request by rstoneback**
Improves input handling for datetime parameters that are more precise than just year, month, and day, where appropriate. As part of this process, I created a date attribute, which now self filters when set. Use of the _filter_datetime method was expanded. Tests were created, 1) Loading with a more precise date, and 2) Repeating all of the basic test but where the file datetimes are more precise than year, month, and day. Fix for issue #122.

Adds a check to ensure that the datetimes for the primary time index are both unique and monotonically increasing. Fix for issue #140.

### CODE REVIEW

**Comment by rstoneback** [comment]
The SuperDARN data is currently loaded into a DataFrame with non-unique times. To incorporate this pull we need to either 
 - put in an option to disable unique time check. While a hack, this lets superdarn continue. Downside, no motivation for future packages to ensure unique times.
 - convert SuperDARN to xarray, and deal with the changes required to processing for the DMSP, SuperDARN, and DINEOFs project.

**Comment by rstoneback** [comment]
Pinging @aburrell since choices here directly affect you.

**Comment by aburrell** [comment]
I move for it to be xarrayed.  This will be initially annoying for dineofs, but I honestly think we'll sort this out before ion is useable again.

**Comment by rstoneback** [comment]
Hahaha :(

On Feb 21, 2019, at 6:38 AM, Angeline Burrell <notifications@github.com<mailto:notifications@github.com>> wrote:


I move for it to be xarrayed. This will be initially annoying for dineofs, but I honestly think we'll sort this out before ion is useable again.

**Comment by rstoneback** [comment]
I agree with the move to xarray. I do like a second opinion though when making work.

**Comment by rstoneback** [comment]
Changes in another project require a pause on upgrading SuperDARN to xarray. We can either table this to a 2.1.0 milestone, or put in a optional flag for the unique and monotonic time check. This pull does include other bug fixes.

**Comment by rstoneback** [comment]
Move forward on option with DeprecationWarning that it will go away.

File: pysat/_instrument.py [hunk 1]
```diff
@@ -609,18 +614,20 @@ def _empty(self, data=None):
...
         """Returns time index of loaded data."""
-
-        if self.pandas_format:
-            return self.data.index
-        else:
-            if 'time' in self.data.indexes:
-                return self.data.indexes['time']
-            else:
-                return pds.Index([])
+        return self._index()
```
Comment (rstoneback) [inline pysat/_instrument.py]: Not sure what I did here.

Comment (rstoneback) [reply pysat/_instrument.py]: Things were good. I sorted it out.

**Comment by rstoneback** [comment]
I think I got that merge correct! :)

**Comment by rstoneback** [comment]
Requesting review from both @jklenzing and @aburrell since Jeff has made most of the changes I merged into and Angeline asked for the strict times feature. Cheers.

File: pysat/tests/test_instrument.py [hunk 2]
```diff
@@ -411,14 +482,50 @@ def test_right_bounds_with_next(self):
...
+
+    def test_set_bounds_with_frequency(self):
+        start = pysat.datetime(2009,1,1)
+        stop = pysat.datetime(2010,1,15)
+        self.testInst.bounds = (start, stop, 'M')
+        assert np.all(self.testInst._iter_list == pds.date_range(start, stop, 
+                                                                 freq='M').tolist())
+                                                                                                                                                                                                                                
+    @raises(ValueError)
+    def test_set_bounds_too_few(self):
```
Comment (jklenzing) [inline pysat/tests/test_instrument.py]: What about a "bounds_too_many" test with three date inputs?

File: pysat/_instrument.py [hunk 3]
```diff
@@ -1260,16 +1279,32 @@ def load(self, yr=None, doy=None, date=None, fname=None, fid=None,
             self.date = pds.datetime(temp.year, temp.month, temp.day)
             self.yr, self.doy = utils.time.getyrdoy(self.date)
 
+        # ensure data is unique and monotonic
+        # check occurs after all the data padding loads, or individual load
+        # thus it can potentially check issues with padding or with raw data
+        if strict_time_flag:
```
Comment (jklenzing) [inline pysat/_instrument.py]: Should this be self.strict_time_flag?

File: pysat/tests/test_instrument.py [hunk 4]
```diff
@@ -564,7 +712,81 @@ def teardown(self):
...
 
+#------------------------------------------------------------------------------
+#
+# Repeat TestBasics above with shifted file dates
+#
+#------------------------------------------------------------------------------
+
+class TestBasicsShiftedFileDates(TestBasics):
+    def setup(self):
+        reload(pysat.instruments.pysat_testing)
```
Comment (jklenzing) [inline pysat/tests/test_instrument.py]: use `re_load` here (see Lines 12-15 for version-dependent import statement)

**Comment by jklenzing** [comment]
Looking good!

File: pysat/tests/test_instrument.py (hunk 2 above)
Comment (rstoneback) [reply pysat/tests/test_instrument.py]: Added.

File: pysat/_instrument.py (hunk 3 above)
Comment (rstoneback) [reply pysat/_instrument.py]: Absolutely.

File: pysat/tests/test_instrument.py (hunk 4 above)
Comment (rstoneback) [reply pysat/tests/test_instrument.py]: Fixed.

File: pysat/_instrument.py [hunk 5]
```diff
@@ -242,8 +254,9 @@ def __init__(self, platform=None, name=None, tag=None, sat_id=None,
         else:
             self._null_data = xr.Dataset(None)
             self._data_library = xr.Dataset
-
+        # assiign null data for user selected data type
```
Comment (jklenzing) [inline pysat/_instrument.py]: spelling

File: pysat/tests/test_instrument.py [hunk 6]
```diff
@@ -564,7 +719,83 @@ def teardown(self):
...
+#
+# Repeat TestBasics above with shifted file dates
+#
+#------------------------------------------------------------------------------
+
+class TestBasicsShiftedFileDates(TestBasics):
+    def setup(self):
+        re_load(pysat.instruments.pysat_testing)
+        """Runs before every method to create a clean testing setup."""
+        self.testInst = pysat.Instrument('pysat', 'testing', '10',
```
Comment (jklenzing) [inline pysat/tests/test_instrument.py]: This functionality changed in #203.
Should be sat_id='10'

File: pysat/tests/test_instrument.py [hunk 7]
```diff
@@ -564,7 +719,83 @@ def teardown(self):
...
+# Test Instrument with a non-unique and non-monotonic index
+#
+#------------------------------------------------------------------------------
+
+
+class TestMalformedIndex():
+    def setup(self):
+        re_load(pysat.instruments.pysat_testing)
+        """Runs before every method to create a clean testing setup."""
+        self.testInst = pysat.Instrument('pysat', 'testing', '10',
```
Comment (jklenzing) [inline pysat/tests/test_instrument.py]: same here

File: pysat/tests/test_instrument.py [hunk 8]
```diff
@@ -564,7 +719,83 @@ def teardown(self):
...
+#
+# Repeat tests above with xarray data
+#
+#------------------------------------------------------------------------------
+
+class TestMalformedIndexXarray(TestMalformedIndex):
+    def setup(self):
+        re_load(pysat.instruments.pysat_testing)
+        """Runs before every method to create a clean testing setup."""
+        self.testInst = pysat.Instrument('pysat', 'testing_xarray', '10',
```
Comment (jklenzing) [inline pysat/tests/test_instrument.py]: again

**Comment by rstoneback** [comment]
The only test failures still left are from SuperMAG's download changes. I can't seem to merge any other branch into this one for some reason. So the test failures will remain. Otherwise this pull should be good to go.

**Pull request merged**